README.md
# Fleet

[![Latest Version on Packagist](https://img.shields.io/packagist/v/jezzdk/fleet-cli.svg?style=flat-square)](https://packagist.org/packages/jezzdk/fleet-cli)
[![Total Downloads](https://img.shields.io/packagist/dt/jezzdk/fleet-cli.svg?style=flat-square)](https://packagist.org/packages/jezzdk/fleet-cli)

Easily run multiple Laravel Sail applications at the same time locally with custom domain names.

> Note: This is still new and may contain bugs, if you see something weird please [create an issue](https://github.com/jezzdk/fleet-cli/issues/new)

- [Installation](#installation)
- [Getting Started](#getting-started)
- [Local SSL](#local-ssl)
- [Port Conflicts](#port-conflicts)
- [Additional Usage](#additional-usage)

## Installation

You can install the package via composer:

```bash
composer require global jezzdk/fleet-cli
```

## Getting Started

Navigate to the root directory of your Laravel application, and stop any Sail instance if it's running. Then, use the following command to add Fleet support to your app:

```bash
fleet add
```

During setup you'll be prompted to enter in a domain name. Alternatively, you can pass it in through the command line:

```bash
fleet add my-app.localhost
```

After the setup finishes, you can start up Sail just like normal:

```bash
./vendor/bin/sail up
```

And your site will be available at the domain you provided!

> Note: If you chose a domain that doesn't end in `.localhost`, you will need to add an entry to your hosts file to direct traffic to 127.0.0.1

## Local SSL

Fleet supports local SSL on your custom domains through the power of [mkcert](https://mkcert.dev). After you've installed it on your machine, you can use the `--ssl` option when using the `fleet add` command to enable it for your application.

```bash
fleet add my-app.localhost --ssl
```

A local certificate will be generated and stored in `~/.config/mkcert/certs`. After spinning up your site with Sail, your specified domain will have https enabled.

## Port Conflicts

When spinning up multiple Laravel Sail appliactions, it's likely you'll encounter an error about port conflicts between Docker containers. This is because each service has a port mapped to your local machine, and by default, they're the same across your applications.

In order to fix this, add different forwarded port numbers to each Laravel application using the `.env` file. For example:

- App #1

```env
FORWARD_DB_PORT=3306
FORWARD_REDIS_PORT=6379
```

- App #2

```env
FORWARD_DB_PORT=4306
FORWARD_REDIS_PORT=7379
```

This way, both applications can be spun up using Fleet and Sail, and their respective services' ports won't conflict.

## Additional Usage

By default, whenever you use `fleet add`, a Docker network and container are both started to handle the traffic from your local domain name(s).

You can start this manually by using:

```bash
fleet start
```

If you would like to remove Fleet support from an application and return it back to the default Docker setup, you can run:

```bash
fleet remove
```

To stop and remove all Fleet containers and networks that are currently running on your system, you can use the command:

```bash
fleet stop
```

## Why use this?

[Laravel Sail](https://laravel.com/docs/sail) uses Docker and Docker Compose to spin up containers that create a local development environment for your application.

By default, the containers are bound to the `:80` port of your local machine. Spinning up a second application results in a failure due to port conflicts, but you can always adjust it so that the second app is available at something like `:8081`.

This can have some unintended consequences though, and can get messy juggling communication between two different applications using port numbers.

Instead, Fleet provides a small set of commands that alter your `docker-compose.yml` file to provide support for [Traefik](https://hub.docker.com/_/traefik), a reverse proxy that runs on a Docker container.

When you add a site to Fleet, a network and a few labels are added to the Docker Compose entry for your application, and a main Traefik container is spun up to handle all local web traffic incoming to the `:80` port.

This configuration allows two or more Laravel Sail applications configured to different domains to resolve to their respective running containers.

For a more in-depth look at how this all ties together, check out [this video](https://www.youtube.com/watch?v=mZbLvGQqEvY) that I published on using Traefik with Docker Compose.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what's recently changed.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `jezzdk/fleet-cli` will be documented in this file.

## 1.0.0 - 2023-05-25

- Initial release!
- Business logic was copied almost as-is from the [original package](https://github.com/aschmelyun/fleet) by [aschmelyun](https://github.com/aschmelyun)


app/Commands/AddCommand.php
<?php

namespace App\Commands;

use App\Fleet;
use App\Support\Filesystem;
use Illuminate\Console\Command;
use Illuminate\Support\Arr;
use Symfony\Component\Yaml\Yaml;

class AddCommand extends Command
{
    public $signature = 'add
                        {domain? : The test domain to use}
                        {--ssl : Include local SSL with mkcert}';

    public $description = 'Installs Fleet support onto the current application';

    public function handle(Filesystem $filesystem): int
    {
        // set the domain to the one the user provided, or ask what it should be
        $domain = $this->argument('domain');
        if (! $domain) {
            $domain = $this->ask('What domain name would you like to use for this app?', 'laravel.localhost');
        }

        $domainArray = explode('.', $domain);
        if (end($domainArray) !== 'localhost') {
            $this->line(' Just a quick note: You may need to update your hosts file to direct traffic to this domain. You can prevent this by using a domain that ends in `.localhost`');
        }

        // determine if laravel/sail is a required-dev package in the root composer file
        if (! $this->isSailInstalled()) {
            $this->error(' Laravel Sail is required for this package');
            $this->line(' For more information, check out https://laravel.com/docs/sail#installation');

            return self::FAILURE;
        }

        // if the docker-compose.yml file isn't available, publish it
        if (! file_exists('docker-compose.yml') && ! file_exists('docker-compose.backup.yml')) {
            $this->info('No docker-compose.yml file available, running sail:install...');
            Fleet::process('php artisan sail:install');
        }

        // copy the docker-compose.yml file for a backup
        if (file_exists('docker-compose.yml')) {
            if ($this->confirm('This will modify your docker-compose.yml file, do you want to back it up first?')) {
                rename('docker-compose.yml', 'docker-compose.backup.yml');
            }
        }

        // determine what port 8081+ is available and set it as the APP_PORT
        $port = 8081;
        while ($this->isPortTaken($port)) {
            $port++;
        }

        try {
            $filesystem->writeToEnvFile('APP_PORT', $port);
        } catch (\Exception $e) {
            $this->error($e->getMessage());

            return self::FAILURE;
        }

        // add a modified docker-compose.yml file to include traefik labels
        $file = 'docker-compose.backup.yml';
        if (! file_exists($file)) {
            $file = 'docker-compose.yml';
        }

        if (! file_exists($file)) {
            $this->error('A docker-compose.yml file or a docker-compose.backup.yml file does not exist');

            return self::FAILURE;
        }

        $yaml = $this->generateYamlForDockerCompose($file, $domain);

        // determine if the user wants to use SSL and add support if so
        if ($this->option('ssl')) {
            $this->info(' 🔒 Adding SSL support...');

            try {
                $filesystem->createCertificates($domain);
            } catch (\Exception $e) {
                $this->error($e->getMessage());
                $this->line('For more information, try checking out the documentation at https://mkcert.dev');

                return self::FAILURE;
            }

            try {
                $filesystem->createSslConfig($domain);
            } catch (\Exception $e) {
                $this->error($e->getMessage());

                return self::FAILURE;
            }

            $heading = str_replace('.', '-', $domain);
            $yaml['services'][$heading]['labels'][] = "traefik.http.routers.{$heading}.tls=true";
        }

        file_put_contents('docker-compose.yml', Yaml::dump($yaml, 6));

        // call fleet:start to determine if the fleet network and traefik container is up
        $this->call('start');

        // return info back to the user
        $this->info(' ✨ All done! You can now run `./vendor/bin/sail up`');
        $this->newLine();

        return self::SUCCESS;
    }

    private function generateYamlForDockerCompose(string $file, string $domain): array
    {
        $yaml = Yaml::parseFile($file);

        $heading = str_replace('.', '-', $domain);

        // resets the top service key to the domain name
        $service = $yaml['services'][array_keys($yaml['services'])[0]];
        unset($yaml['services'][array_keys($yaml['services'])[0]]);

        // adds the entire services array back with the new domain key
        $yaml['services'] = [$heading => $service, ...$yaml['services']];

        // adds the traefik labels to the yaml file
        $yaml['services'][$heading]['networks'][] = 'fleet';
        $yaml['services'][$heading]['labels'] = [
            "traefik.http.routers.{$heading}.rule=Host(`{$domain}`)",
            "traefik.http.services.{$heading}.loadbalancer.server.port=80",
        ];

        // removes port binding for our app service
        unset($yaml['services'][$heading]['ports'][0]);
        $yaml['services'][$heading]['ports'] = array_values($yaml['services'][$heading]['ports']);

        // adds the fleet network
        $yaml['networks']['fleet']['external'] = true;

        return $yaml;
    }

    private function isPortTaken($port): bool
    {
        $process = Fleet::process("lsof -nP -iTCP:{$port} -sTCP:LISTEN");

        return boolval($process->getOutput());
    }

    private function isSailInstalled(): bool
    {
        if (! file_exists('composer.json')) {
            return false;
        }

        $composer = json_decode(file_get_contents('composer.json'), true);

        return Arr::has($composer, 'require-dev.laravel/sail');
    }
}
